In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

%matplotlib inline

In [9]:
import pickle 

# Define the filename of the pickled file filename = 'data.pickle' 
# Unpickle the object from the file with open(filename, 'rb') as f: unpickled_object = pickle.load(f) 
# Print the unpickled object print(unpickled_object)

filename = 'control_confirmed.pkl'
with open(filename, 'rb') as f: 
    control_confirmed = pickle.load(f) 
control_confirmed

0  1
0      4033851  3
1      6651403  1
2      2105948  1
3      9814849  1
4      4446011  1
...        ... ..
15429  1011862  4
15430  8930684  1
15431  4840338  3
15432  4210290  1
15433  1684561  1

[15434 rows x 2 columns]

In [10]:
control_confirmed.rename(columns = {0:'client_id', 1:'count_of_steps'}, inplace=True)
control_confirmed

client_id  count_of_steps
0       4033851               3
1       6651403               1
2       2105948               1
3       9814849               1
4       4446011               1
...         ...             ...
15429   1011862               4
15430   8930684               1
15431   4840338               3
15432   4210290               1
15433   1684561               1

[15434 rows x 2 columns]

In [11]:
filename = 'how_long_in_step.pkl'
with open(filename, 'rb') as f: 
    time_on_steps = pickle.load(f) 
time_on_steps

0        1  2                   3  \
0        68804396_74737424429_397490  1000062  0 2017-04-06 15:38:49   
1        68804396_74737424429_397490  1000062  1 2017-04-06 15:41:51   
2        68804396_74737424429_397490  1000062  2 2017-04-06 15:42:01   
3        68804396_74737424429_397490  1000062  3 2017-04-06 15:42:50   
4        68804396_74737424429_397490  1000062  4 2017-04-06 15:54:49   
...                              ...      ... ..                 ...   
320220  870243567_56915814033_814203  9999729  4 2017-05-08 16:09:40   
320221   99583652_41711450505_426179  9999729  0 2017-04-05 13:40:49   
320222   99583652_41711450505_426179  9999729  1 2017-04-05 13:41:04   
320223  472154369_16714624241_585315  9999832  0 2017-05-16 16:46:03   
320224  472154369_16714624241_585315  9999832  1 2017-05-16 16:46:11   

                         4      5        6  
0      2017-04-06 15:41:51  182.0  Control  
1      2017-04-06 15:42:01   10.0  Control  
2      2017-04-06 15:42:50   49.0  Control  
3      2017-04-06 15:54:49  719.0  Control  
4                      NaT    NaN  Control  
...                    ...    ...      ...  
320220                 NaT    NaN     Test  
320221 2017-04-05 13:41:04   15.0     Test  
320222                 NaT    NaN     Test  
320223 2017-05-16 16:46:11    8.0     Test  
320224                 NaT    NaN     Test  

[320225 rows x 7 columns]

In [12]:
#rename columns to be descriptive

time_on_steps.rename(columns = {0:'visit_id',1:'client_id',2:'process_step',3:'date_time',4:'next_step_time', 5:'time_spent_sec', 6:'Variation'},inplace=True)
time_on_steps

visit_id client_id process_step  \
0        68804396_74737424429_397490   1000062            0   
1        68804396_74737424429_397490   1000062            1   
2        68804396_74737424429_397490   1000062            2   
3        68804396_74737424429_397490   1000062            3   
4        68804396_74737424429_397490   1000062            4   
...                              ...       ...          ...   
320220  870243567_56915814033_814203   9999729            4   
320221   99583652_41711450505_426179   9999729            0   
320222   99583652_41711450505_426179   9999729            1   
320223  472154369_16714624241_585315   9999832            0   
320224  472154369_16714624241_585315   9999832            1   

                 date_time      next_step_time  time_spent_sec Variation  
0      2017-04-06 15:38:49 2017-04-06 15:41:51           182.0   Control  
1      2017-04-06 15:41:51 2017-04-06 15:42:01            10.0   Control  
2      2017-04-06 15:42:01 2017-04-06 15:42:50            49.0   Control  
3      2017-04-06 15:42:50 2017-04-06 15:54:49           719.0   Control  
4      2017-04-06 15:54:49                 NaT             NaN   Control  
...                    ...                 ...             ...       ...  
320220 2017-05-08 16:09:40                 NaT             NaN      Test  
320221 2017-04-05 13:40:49 2017-04-05 13:41:04            15.0      Test  
320222 2017-04-05 13:41:04                 NaT             NaN      Test  
320223 2017-05-16 16:46:03 2017-05-16 16:46:11             8.0      Test  
320224 2017-05-16 16:46:11                 NaT             NaN      Test  

[320225 rows x 7 columns]

In [23]:
# Step 1: Filter DataFrame for 'start' steps
start_step = time_on_steps[time_on_steps['process_step'] == '0']

# Step 2: Group by 'client_id'
grouped = start_step.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_start = grouped['time_spent_sec'].mean()

average_time_start = pd.DataFrame(average_time_spent_start).reset_index()
average_time_start

client_id  time_spent_sec
0       1000062      182.000000
1       1000353       36.666667
2       1000536        4.500000
3       1000672       22.000000
4       1001101       10.000000
...         ...             ...
50071   9999150        6.000000
50072   9999400       22.000000
50073   9999626        8.000000
50074   9999729       34.000000
50075   9999832        8.000000

[50076 rows x 2 columns]

In [24]:
# Step 1: Filter DataFrame for 'start' steps
step_1 = time_on_steps[time_on_steps['process_step'] == '1']

# Step 2: Group by 'client_id'
grouped = step_1.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_1 = grouped['time_spent_sec'].mean()

average_time_step_1 = pd.DataFrame(average_time_spent_1).reset_index()
average_time_step_1

client_id  time_spent_sec
0       1000062            10.0
1       1000353            46.0
2       1000536             3.0
3       1000672            35.0
4       1001101            28.5
...         ...             ...
44414   9999150            12.0
44415   9999400            27.0
44416   9999626             NaN
44417   9999729           148.5
44418   9999832             NaN

[44419 rows x 2 columns]

In [26]:
# Step 1: Filter DataFrame for 'start' steps
step_2 = time_on_steps[time_on_steps['process_step'] == '2']

# Step 2: Group by 'client_id'
grouped = step_2.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_2 = grouped['time_spent_sec'].mean()

average_time_step_2 = pd.DataFrame(average_time_spent_2).reset_index()
average_time_step_2

client_id  time_spent_sec
0       1000062           49.00
1       1000353          162.00
2       1000536           85.00
3       1000672           24.00
4       1001101          102.75
...         ...             ...
40903   9998851          101.00
40904   9998964          107.00
40905   9999009           19.00
40906   9999400           46.00
40907   9999729           60.00

[40908 rows x 2 columns]

In [27]:
# Step 1: Filter DataFrame for 'start' steps
step_3 = time_on_steps[time_on_steps['process_step'] == '3']

# Step 2: Group by 'client_id'
grouped = step_3.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_3 = grouped['time_spent_sec'].mean()

average_time_step_3 = pd.DataFrame(average_time_spent_3).reset_index()
average_time_step_3

client_id  time_spent_sec
0       1000062           719.0
1       1000353            52.0
2       1000536            16.0
3       1000672            58.0
4       1001101            81.0
...         ...             ...
38298   9998851            31.0
38299   9998964            74.0
38300   9999009           118.0
38301   9999400            24.0
38302   9999729            21.0

[38303 rows x 2 columns]

In [29]:
time_spent_on_confirm = time_on_steps[time_on_steps['process_step'] == '4'][['client_id','process_step','time_spent_sec']]
time_spent_on_confirm

client_id process_step  time_spent_sec
4        1000062            4             NaN
11       1000353            4             NaN
18       1000536            4             NaN
23       1000672            4             NaN
28       1001101            4             NaN
...          ...          ...             ...
320190   9998851            4             NaN
320196   9998964            4             NaN
320201   9999009            4             NaN
320209   9999400            4             NaN
320220   9999729            4             NaN

[43082 rows x 3 columns]

In [31]:
time_spent_on_confirm['client_id'].duplicated().sum()

8961

In [28]:
# Step 1: Filter DataFrame for 'start' steps
confirm_step = time_on_steps[time_on_steps['process_step'] == '4']

# Step 2: Group by 'client_id'
grouped = confirm_step.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_confirm = grouped['time_spent_sec'].mean()

average_time_step_confirm = pd.DataFrame(average_time_spent_confirm).reset_index()
average_time_step_confirm

client_id  time_spent_sec
0       1000062             NaN
1       1000353             NaN
2       1000536             NaN
3       1000672             NaN
4       1001101             NaN
...         ...             ...
34116   9998851             NaN
34117   9998964             NaN
34118   9999009             NaN
34119   9999400             NaN
34120   9999729             NaN

[34121 rows x 2 columns]

In [ ]:
#No time spent on 4 because that's when the completed the process... duh!

Started 50,076

Confirmed 34,121

15,955 clients abandoned the process

(15955/50076)*100 = 31.9% (abandonment rate)